Compiler hang with incomplete struct literal at end of file

This malformed code ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=92c9932ec918d7af092fa140815f3e0a)) hangs the compiler:

```rust
struct R { }

struct S {
    x: [u8; R
```

Found by fuzzing with a [modified fuzz-rustc](https://github.com/jruderman/fuzz-rustc).

### The bug's edges
* `R` must be defined as a named-field struct (not a tuple struct or unit struct), although it can have 0 fields.
* `R` must be the last token to produce the hang:
  * No opening brace (note that braces are required for literals of named-field structs)
  * No closing delimiters
  * Not even a comment

### Where it hangs
If I understand [mj.sample.txt](https://github.com/rust-lang/rust/files/9847560/mj.sample.txt) correctly, the hang occurs in LateResolutionVisitor, with `try_lookup_name_relaxed` never finishing (while calling various functions such as  `smart_resolve_context_dependent_help` and `span_to_snippet`)

### Additional example

This also hangs. I'm guessing it's the same bug.

```rust
struct R { }

fn f() {
    R
```

### Version

`rustc --version --verbose`:
```
rustc 1.66.0-nightly (6e95b6da8 2022-10-22)
binary: rustc
commit-hash: 6e95b6da885f42a4e1314595089fa4295e329d11
commit-date: 2022-10-22
host: x86_64-apple-darwin
release: 1.66.0-nightly
LLVM version: 15.0.2
```

### Regression

[Bisected](https://github.com/rust-lang/cargo-bisect-rustc) with: 
```bash
cargo-bisect-rustc --start=2022-10-03 --end=nightly-2022-10-23 --preserve --script=./timeout.sh
```

<details>
<summary>timeout.sh</summary>
<p>

```bash
#!/bin/bash

# This script is wrong in several ways:
# - It will try to kill processes that might no longer exist
# - It may leave an orphan rustc process after the bisect (because it only kills the shell, not rustc)
# But it seems to be good enough for bisection to work

killall rustc
outershell=$$
(sleep 5; kill $outershell) &
rustc src/main.rs
true
```

</p>
</details>

* regressed nightly: nightly-2022-10-21
* regressed commit: https://github.com/rust-lang/rust/commit/53728ff751df4c271d4ea565b6871057a3504fc5 (PR #103185)

@rustbot label +regression-from-stable-to-nightly +I-hang 

<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_DATA_START$${"user":"chenyukang"}$$TRIAGEBOT_ASSIGN_DATA_END -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

I can not reproduce it with latest code, did I missed anything?

```console
cat@LAPTOP-V6U0QKD4:~/code/rust$ cat hang.rs
struct R { }

struct S {
    x: [u8; R
cat@LAPTOP-V6U0QKD4:~/code/rust$ rustc -vV
rustc 1.66.0-nightly (7fcf850d7 2022-10-23)
binary: rustc
commit-hash: 7fcf850d7942804990a1d2e3fe036622a0fe4c74
commit-date: 2022-10-23
host: x86_64-unknown-linux-gnu
release: 1.66.0-nightly
LLVM version: 15.0.2
cat@LAPTOP-V6U0QKD4:~/code/rust$ rustc hang.rs
error: this file contains an unclosed delimiter
 --> hang.rs:4:15
  |
3 | struct S {
  |          - unclosed delimiter
4 |     x: [u8; R
  |        -      ^
  |        |
  |        unclosed delimiter

error[E0423]: expected value, found struct `R`
 --> hang.rs:4:13
  |
1 | struct R { }
  | ------------ `R` defined here
...
4 |     x: [u8; R
  |             ^ help: use struct literal syntax instead: `R {}`

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0423`.
```

--- a/compiler/rustc_span/src/source_map.rs
+++ b/compiler/rustc_span/src/source_map.rs
@@ -855,7 +855,8 @@ impl SourceMap {
     /// Returns a new span representing the next character after the end-point of this span.
     /// Special cases:
     /// - if span is a dummy one, returns the same span
-    /// - if next_point reached the end of source, return span with lo = hi
+    /// - if next_point reached the end of source, return a span exceeding the end of source,
+    ///   which means sm.span_to_snippet(next_point) will get `Err`
     /// - respect multi-byte characters
     pub fn next_point(&self, sp: Span) -> Span {
         if sp.is_dummy() {
@@ -864,9 +865,6 @@ impl SourceMap {
         let start_of_next_point = sp.hi().0;
 
         let width = self.find_width_of_character_at_span(sp, true);
-        if width == 0 {
-            return Span::new(sp.hi(), sp.hi(), sp.ctxt(), None);
-        }
         // If the width is 1, then the next span should only contain the next char besides current ending.
         // However, in the case of a multibyte character, where the width != 1, the next span should
         // span multiple bytes to include the whole character.
@@ -938,7 +936,7 @@ impl SourceMap {
         // Ensure indexes are also not malformed.
         if start_index > end_index || end_index > source_len - 1 {
             debug!("find_width_of_character_at_span: source indexes are malformed");
-            return 0;
+            return 1;
         }
 
         let src = local_begin.sf.external_src.borrow();

--- a/compiler/rustc_span/src/source_map/tests.rs
+++ b/compiler/rustc_span/src/source_map/tests.rs
@@ -511,16 +511,17 @@ fn test_next_point() {
     assert_eq!(span.lo().0, 4);
     assert_eq!(span.hi().0, 5);
 
-    // A non-empty span at the last byte should advance to create an empty
-    // span pointing at the end of the file.
+    // Reaching to the end of file, return a span that will get error with `span_to_snippet`
     let span = Span::with_root_ctxt(BytePos(4), BytePos(5));
     let span = sm.next_point(span);
     assert_eq!(span.lo().0, 5);
-    assert_eq!(span.hi().0, 5);
+    assert_eq!(span.hi().0, 6);
+    assert!(sm.span_to_snippet(span).is_err());
 
-    // Empty span pointing just past the last byte.
+    // Reaching to the end of file, return a span that will get error with `span_to_snippet`
     let span = Span::with_root_ctxt(BytePos(5), BytePos(5));
     let span = sm.next_point(span);
     assert_eq!(span.lo().0, 5);
-    assert_eq!(span.hi().0, 5);
+    assert_eq!(span.hi().0, 6);
+    assert!(sm.span_to_snippet(span).is_err());
 }

--- a/src/test/ui/parser/issue-103451.rs
+++ b/src/test/ui/parser/issue-103451.rs
@@ -1,0 +1,5 @@
+// error-pattern: this file contains an unclosed delimiter
+// error-pattern: expected value, found struct `R`
+struct R { }
+struct S {
+    x: [u8; R

--- a/src/test/ui/parser/issue-103451.stderr
+++ b/src/test/ui/parser/issue-103451.stderr
@@ -1,0 +1,32 @@
+error: this file contains an unclosed delimiter
+  --> $DIR/issue-103451.rs:5:15
+   |
+LL | struct S {
+   |          - unclosed delimiter
+LL |     x: [u8; R
+   |        -      ^
+   |        |
+   |        unclosed delimiter
+
+error: this file contains an unclosed delimiter
+  --> $DIR/issue-103451.rs:5:15
+   |
+LL | struct S {
+   |          - unclosed delimiter
+LL |     x: [u8; R
+   |        -      ^
+   |        |
+   |        unclosed delimiter
+
+error[E0423]: expected value, found struct `R`
+  --> $DIR/issue-103451.rs:5:13
+   |
+LL | struct R { }
+   | ------------ `R` defined here
+LL | struct S {
+LL |     x: [u8; R
+   |             ^ help: use struct literal syntax instead: `R {}`
+
+error: aborting due to 3 previous errors
+
+For more information about this error, try `rustc --explain E0423`.

